### PR TITLE
tried to implement my vision of arrayList

### PR DESCRIPTION
**Pull request by happytroya**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,135 @@
...
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_RATE = 1.5;
+    private static final Object[] EMPTY_ELEMENT_DATA = {};
+    private Object[] elementData;
+    private int size = 0;
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:11]:
```suggestion
    private int size;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,135 @@
...
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+
+        if (size == elementData.length) {
+            growStorage();
+        }
 
+        for (int i = size - 1; i >= index; i--) {
+            elementData[i + 1] = elementData[i];
+        }
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:47]: use System.arrayCopy to move elements, you dont need another array for this

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,135 @@
...
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        final Object[] tmp = elementData;
+
+        final T oldValue = (T)tmp[index];
+
+        for (int i = index; i < size - 1; i++) {
+            elementData[i] = elementData[i + 1];
+        }
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:91]: use System.arrayCopy, you dont need additional array to copy values. Just copy them to the same array

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,135 @@
...
+                    return remove(i);
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (elementData[i] != null && elementData[i].equals(element)) {
+                    return remove(i);
+                }
+            }
+        }
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:112]: identical blocks, combine those ifs
for (...) {
    if (condition1 || condition2) {
    }
}

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java (hunk 4 above)
Comment (happytroya) [reply src/main/java/core/basesyntax/ArrayList.java:112]: Can I use ArrayUtils.add instead of System.arrayCopy?

Comment (kozhukhovsky) [reply src/main/java/core/basesyntax/ArrayList.java]: no :)

**Review by kozhukhovsky — CHANGES_REQUESTED** [review]
ok, but still need to use System.arrayCopy :)

**Comment by happytroya** [comment]
Why? It wasn't mentioned in task, so I'm interested int this question

**Pull request closed without merging**